### ISSUE TITLE
Feature request: Support FlexSearch 0.7.x

### ISSUE THREAD

**Issue by haysclark** [angeloashmore/gatsby-plugin-local-search#44]
new version of [FlexSearch 0.7.0](https://github.com/nextapps-de/flexsearch/releases/tag/0.7.0) that has some breaking changes with FlexSearch 0.6.x.

**Comment by haysclark** [comment]
It might be worth moving `flexsearch` to _peerDependencies_... Maybe that should be a separate Issue that should be resolved prior to FlexSearch 0.7.x support?

Current _gatsby-plugin-local-search_ [package.json](https://github.com/angeloashmore/gatsby-plugin-local-search/blob/v2.0.1/package.json#L40-L48)

```json
...
  "dependencies": {
    "flexsearch": "^0.6.32",
  },
  "peerDependencies": { 
   "gatsby": "&gt;= 2.20.0"
  },
...
```

**Comment by haysclark** [comment]
My first thought was, "Move 'flexsearch' to peerDependencies". Unfortunately, this would result in users having to install both _gatsby-plugin-local-search_ and _flexsearch_ even if they are NOT using _flexsearch_ for local search.

Maybe it would be worth the effort to upgrade the library to be more modular and have the 'engines' broken out into sub-plugin? Sorta like the structure of the [gatsby-transformer-remark](https://www.gatsbyjs.com/plugins/gatsby-transformer-remark/) ecosystem.

proposed _gatsby-config.js_:

```javascript
module.exports = {
  plugins: [
    {
      resolve: "gatsby-plugin-local-search",
      options: {
        // engine: "flexsearch", <-- omit and rely on plugins to define engine?
        name: "pages",
        // new section
        plugins: [
          `gatsby-local-search-flexsearch`,
          // OR
          {
            resolve: `gatsby-local-search-flexsearch`,
            options: {
              // ... 
            },
          },
        ],
        // end of new section
        query: `
  query {
    allMarkdownRemark(sort: { fields: [frontmatter___date], order: ASC }) {
      nodes {
        id
        fields {
          slug
        }
        frontmatter {
          date(formatString: "MMMM DD, YYYY")
          title
        }
        rawMarkdownBody
      }
    }
  }
`,
        ref: "id",
        index: ["title", "body"],
        store: ["id", "title", "body", "date", "slug", "type"],
        normalizer: ({ data }) =>
          data.allMarkdownRemark.nodes.map(node => ({
           // ... code
          })),
      },
    },
  ],
}
```

**Comment by angeloashmore** [comment]
Hey @haysclark, this is a great recommendation. Splitting into sub-plugins also opens the system up to others search engines more easily.

The `plugins` option is required to be an array, like you have in your example. In the case of this plugin, multiple sub-plugins isn't applicable since it should only produce one index. Is there any need to build a FlexSearch and Lunr index for the same set of data?

Rather than use `plugins` we could use the `engine` option as a way to load the sub-plugin. For example, the config could look something like this:

```js
{
  resolve: 'gatsby-plugin-local-search',
  options: {
    name: 'pages',
    engine: 'gatsby-local-search-flexsearch',
    query: `
      query {
        allMarkdownRemark(sort: { fields: [frontmatter___date], order: ASC }) {
          nodes {
            id
            fields {
              slug
            }
            frontmatter {
              date(formatString: "MMMM DD, YYYY")
              title
            }
            rawMarkdownBody
          }
        }
      }
    `,
    index: ['title', 'body'],
    store: ['id', 'title', 'body', 'date', 'slug', 'type'],
    normalizer: ({ data }) =>
      data.allMarkdownRemark.nodes.map((node) => ({
        // ... code
      }))
  },
}
```

This keeps the API similar to the existing API and allows us to migrate from `engine: 'flexsearch'` to `engine: 'gatsby-local-search-flexsearch'` progressively. For example, we could detect the current "flexsearch" and "lunr" options and warn the user to install the plugin and change the config.

`gatsby-local-search-flexsearch` would be loaded dynamically with its own FlexSearch-specific logic.

What do you think?

**Comment by haysclark** [comment]
@angeloashmore Awesome! Yeah that is a good point about 1:1 relationship as well at the point about backwards compatibility. Just to clarify, are you imagining we can support providing custom options to the child engines?

```javascript
{
  resolve: 'gatsby-plugin-local-search',
  options: {
    name: 'pages',
    // for legacy and when providing no custom options
    engine: 'gatsby-local-search-flexsearch',
    // for providing custom options
    engine: {
      resolve: `gatsby-local-search-flexsearch`,
      options: {
        // ... 
      },
    },
    query: `
      query {
        allMarkdownRemark(sort: { fields: [frontmatter___date], order: ASC }) {
          nodes {
            id
            fields {
              slug
            }
            frontmatter {
              date(formatString: "MMMM DD, YYYY")
              title
            }
            rawMarkdownBody
          }
        }
      }
    `,
    index: ['title', 'body'],
    store: ['id', 'title', 'body', 'date', 'slug', 'type'],
    normalizer: ({ data }) =>
      data.allMarkdownRemark.nodes.map((node) => ({
        // ... code
      }))
  },
}
```

**Comment by angeloashmore** [comment]
Good catch on the engine options. I forgot about that part!

The current API allows for an `engineOptions` value. Currently, it is only provided to the FlexSearch engine and is effectively passed through to the index creator as is. In the approach we're discussing, it would be passed to the engine plugin directly for the plugin to do what it needs with it.

For the sake of API simplicity, I'd lean toward using the separate `engine` and `engineOptions` properties rather than a single `engine` option that takes a string or an object with nested properties. I know your example mimics the `plugins` API, but in this case, I think it may introduce added complexity unnecessarily.

Do you think there would be a benefit to using the `resolve`/`options` form instead?

**Comment by haysclark** [comment]
> Do you think there would be a benefit to using the `resolve`/`options` form instead?

I don't think it is really required, I was just following the Gatsby conventions that I'm used too seeing.

**Comment by haysclark** [comment]
@angeloashmore Hmm.... it looks like FlexSearch 0.7.x's [import]:https://github.com/nextapps-de/flexsearch#import/[export](https://github.com/nextapps-de/flexsearch#export) API's have changed quite a bit. FlexSearch 0.7.x exports to multiple data blobs that have to be imported in separate chunks. 

Additionally, there appears to be a flaw in the new `export()` API, because one can not determine when the exporting process has been completed. https://github.com/nextapps-de/flexsearch/issues/235

**Comment by gabrielcsapo** [comment]
is there still work being done adding 0.7 support?

**Comment by angeloashmore** [comment]
@gabrielcsapo There isn't any active development for 0.7 support, but I would be happy to review a PR. To keep the upgrade simple, we can either go the plugin route described in this issue, or we can keep the existing system while supporting 0.7.

**Comment by davidpaulsson** [comment]
Hey everyone :) Does anyone have a working solution for Gatsby + Flexsearch@7+?
